### ISSUE TITLE
Proof-of-work - what it be? (formerly: To merge or not to merge)

### ISSUE THREAD

**Issue by jwinterm** [wownero/meta#3]
That is the question. 

There's already been some discussion about switching to a unique proof-of-work, either by modifying cryptonight further or by abandoning it altogether, with the intent to provide some security through obscurity from 51% attacks by large pools or nicehash. See open issue here: https://github.com/wownero/wownero/issues/15

However, another option that may provide superior security is moving to merge mining with Monero. This would mean tracking changes from their PoW algorithm, and things may get a little hairy around when they hardfork, but if a significant portion of the Monero network adopted Wownero merge mining, this would provide network security superior to the approach outlined in the previous paragraph. 

This is far from a decided issue, and there are certainly other alternatives - just tracking changes to Monero's PoW without merge mining, or just sticking with cnv7, masternodes operating network, whatever, but this thread is for discussion of particularly merge mining as a possibility.

**Comment by cryptonote-social** [comment]
Hmmm, are there any other coins out there already that do merged mining w/ Monero?

**Comment by jwinterm** [comment]
No, not really. Kind of ducknote/darknote/digitalnote is merge-mineable, and monetaverde was/is merge-mineable, but there has never been an open source pool implementation, so if you want to merge mine you are basically at the mercy of minergate.

**Comment by ghost** [comment]
imo merged mining would advertise this coin a lot! Just like Elastos, its community is expecting a steep growth when merged mining with bitcoin will be launched in december. What would be the drawbacks of merged mining? Only the hardfrork related ones?

**Comment by Cactii1** [comment]
Have to remember that this is also like pegging your country's currency to the USD. Wowneo will never (or very rarely) move independently in price from Monero.

**Comment by ghost** [comment]
Although there might be better network security, one of the problems with merge mining is that it encourages “hash & dump” behavior, so the only reason to mine WOW would be to sell it off quickly, which in the long term, could be detrimental to maintaining an active community. I would prefer to track changes to Monero's PoW and when we have a large enough baseline network hash rate, fork to a unique PoW and call it something like “PoW-WOW”.

**Comment by Gingeropolous** [comment]
I'd prefer wownero push to do extremely bonkers things that really push the envelope. Merge mining seems safe. It doesn't seem wow enough. this seems wow: https://github.com/tevador/RandomJS

**Comment by mattcode55** [comment]
Alternative to @Gingeropolous' suggestion: https://github.com/hyc/randprog

**Comment by jwinterm** [comment]
I think that is an interesting option, but would probably require some
heavy lifting for integration and at least some extended testing. Maybe
this is something we could put a wow non forum funding system bounty
together for, if most folks think this is the way to go...

**Comment by ghost** [comment]
FFS bounty +1

**Comment by imperdin** [comment]
@Gingeropolous I don't see any added value in using RandomJS, while it's an additional ASIC countermeasure it's far from best practice to prevent FPGAs.

It would be most effective both for development & time wise to switch to a slightly tweaked version of Heavy
Which is more resistant to ASICs/FPGAs mining in addition to Bot-nets
And it will not be included In Nicehash as it's a tweaked Heavy version named Crypto-WOW

**Comment by jwinterm** [comment]
@imperdin if by "more resistant to ASICs, FPGAs, and botnets" you mean takes longer to verify with CPU, I'm not sure that's the direction we should move in. I like the idea that people can CPU mine, and I definitely favor faster verification.

**Comment by ghost** [comment]
If hobbyists can build FPGAs with widely available consumer components, I have no issue with them. The main problem are ASICs built exclusively by only a handful of manufacturers. As for botnets, network administrators and Windows users need to take some personal responsibility for maintaining an up to date and secure system.

**Comment by ghost** [comment]
~~Extra brownie points if the PoW can incorporate computation of workunits for the [BOINC project](https://boinc.berkeley.edu). https://boinc.berkeley.edu/trac/wiki/ProjectMain~~ *shower thought*

**Comment by guystyles** [comment]
I second the opinion above that wownero should push the envelope. All or nothing! No regrets! Amaze the world, and make them say WOW

**Comment by tevador** [comment]
https://github.com/tevador/RandomJS/issues/2

All features of the RandomJS generator have been implemented. I need help with testing.

Also the decision should be made if my proposed DoS resistant algorithm is to be implemented (this requires a modification of the block format) or we just replace Cryptonight with RandomJS directly.

**Comment by Gingeropolous** [comment]
@tevador , 

> if my proposed DoS resistant algorithm

You mean the asymmetric algorithm design? that makes sense to me. I haven't heard any negatives about an asymemetric design, and indeed I think its been a common criticism of cryptonight. So I'd be for modifying the block format to make it work.

**Comment by tevador** [comment]
> You mean the asymmetric algorithm design? that makes sense to me. I haven't heard any negatives about an asymemetric design, and indeed I think its been a common criticism of cryptonight. So I'd be for modifying the block format to make it work.

Actually, the DoS resistance and asymmetry are separate options.

1. Asymmetry just restricts what a valid solution is. It only affects mining and validation code, so it doesn't need any additional effort to implement since we are already changing the PoW.
1. The DoS resistance requires additional 32 bytes of data in the block, so it needs more effort to implement (for example, block format, daemon and blockchain database has to be modified).

The DoS resistance is not strictly necessary, but rather a nice-to-have since PoW validation will be roughly as slow as Cryptonight (~10 ms).

**Comment by binaryFate** [comment]
Besides the joke and fun, the interest of Wownero for me is to test crazy things (but crazy things that may one day be incorporated in Monero). 
For instance right now it incorporates bulletproof (so does the Monero testnet) but provides a more interesting incentive to attack it, as contrary to Monero testnet it holds some value. Arguably very little, but still more than zero. 
This makes it a cool test bed to push cutting-edge ideas into production and learn from it.

Hence, I also really would like to see a randomJS approach.

**Comment by jwinterm** [comment]
@tevador can you expand a bit on how the DoS prevention would work.

I'll try to find Windows machine to build and test on this week.

Personally I think it would probably be better to implement PoW change first and then DoS later/separately, if that doesn't cause additional complications, so that we're only breaking one thing at a time.

**Comment by ghost** [comment]
Agree with @jwinterm. If DoS resistance is considered more of an optimization, then tinkering with the new PoW should be looked at first.

**Comment by tevador** [comment]
@jwinterm 
I tried to make a diagram to explain the algorithm.

![diagram](https://raw.githubusercontent.com/tevador/RandomJS/master/doc/diagram.png)

A network node can quickly verify the PoW of a new block by following the green arrows. This requires just two Blake2b hash calculations. Therefore even if an attacker can saturate the network connection with invalid blocks, the node can check and discard all of them.

If the intermediate result is not stored in the block, then the whole algorithm must be run to obtain the PoW hash (just like Cryptonight).

Also this is a good time to review the algorithm if we can find any obvious weak points. Comments are appreciated.

**Comment by binaryFate** [comment]
@tevador Are you going to rewrite it in C++?

**Comment by tevador** [comment]
@binaryFate Yes, after the optimal generator options are found so the code can be better optimized.

**Comment by binaryFate** [comment]
The optimization should be done after so it is based on the performance and timing of the C++ version.

I am one of the authors of this software package to optimize algorithm parameters. It has produced many state-of-the-art algorithms (from a handful of parameters to hundreds of them) in different fields. http://iridia.ulb.ac.be/irace/
Would be easy to plug with your algo and optimize for whatever metric you want. I can help with that.

**Comment by tevador** [comment]
> The optimization should be done after so it is based on the performance and timing of the C++ version.

I'm timing only the program execution and for that I use NodeJS, which already natively includes Chrome V8. I think the overhead is minimal. 

> I am one of the authors of this software package to optimize algorithm parameters. It has produced many state-of-the-art algorithms (from a handful of parameters to hundreds of them) in different fields. http://iridia.ulb.ac.be/irace/
> Would be easy to plug with your algo and optimize for whatever metric you want. I can help with that.

Sounds interesting. I've never heard of iterated race. How do you specify the goals that the optimizer should try to meet? Is there something like a fitness function like in GA?

**Comment by binaryFate** [comment]
> Sounds interesting. I've never heard of iterated race. How do you specify the goals that the optimizer should try to meet? Is there something like a fitness function like in GA?

Yes. You need to return a single number to the optimizer. The program is launched through a wrapper (typically a bash script) which can aggregate all returned values into a single number. For instance a weighted sum of the difference to the target runtime (5ms?) plus an optional penalty if the time varies too much across executions, or the amount of RAM, etc. Up to you.
The iterated race is just a very quick effective way to scrap bad candidates and sample new ones from promising ones, it converges faster than what typical GA would do for algorithm configuration problems. It samples around several promising candidates so you focus on several local optima at the same time. The software is designed to sample new candidates differently depending on the type of parameters. For instance if you have a discrete parameter (whose value can be "use_component_a", "use_component_b", "use_component_c"), the values will be sampled differently than if a parameter is an integer or has any logical progresion (say, possible values "high", "medium", "low"), or if it's just a float within [X,Y].
In practice, you write a simple definition of the parameters types and ranges in a text file, your experimental budget, and it does the rest to give you in the end a handful of winning configurations for your particular fitness function.

**Comment by tevador** [comment]
@binaryFate Perfect, that's exactly what we need. RandomJS currently has 113 configurable parameters. I can prepare a script for the optimization. Is there some specification available what format to expect on input/output?

**Comment by cryptonote-social** [comment]
Pretty risky move I think; this is a significant departure from existing PoW approaches that have proven hardness properties.  It's likely there will be clever folks who can implement dramatically faster Javascript engines for these sorts of simple programs than whatever the regular miners will rely on.

**Comment by jwinterm** [comment]
Are there really more clever folks than the google javascript folks? What exactly is the simplification beyond total google js engine?

**Comment by tevador** [comment]
> Pretty risky move I think; this is a significant departure from existing PoW approaches that have proven hardness properties.

Every single existing PoW approach turned out to be susceptible to ASIC implementation. In my opinion, we need to try something radically different.

> It's likely there will be clever folks who can implement dramatically faster Javascript engines for these sorts of simple programs than whatever the regular miners will rely on.

There might be some room for improvements, but something dramatical is unlikely, The randomly generated program is [not so simple](https://pastebin.com/4JPV9HU2). I tried to use as many features of the language as possible. Actually it's an impressive performance from the V8 to parse, compile and execute 40 kilobytes of javascript source code in under 10 milliseconds.

**Comment by cryptonote-social** [comment]
"Are there really more clever folks than the google javascript folks? What exactly is the simplification beyond total google js engine?"

The Google javascript folks, the Apple javascript folks, the Mozilla ones, all have been optimizing Javascript for actual web workloads.  There are undoubtedly optimizations they either have yet to pursue or have abandoned because they weren't worth it for this use case, but might very well apply to these artificial programs.   Even the Cuckoo cycle folks were surprised by some of the optimizations folks came up with for their very simple problem.  They were smart about this though and offered big bounties for faster implementations before advocating it be adopted.  If you are to go this route, you should do the same.  I will bet you a good chunk of $TRTL that someone will find something faster!

**Comment by guystyles** [comment]
What’s the worse thing that can happen if someone does?

**Comment by cryptonote-social** [comment]
_"What’s the worse thing that can happen if someone does?"_

51% attack.

Though most likely outcome is one (or a few) miners making bank at the expense of everyone else.

**Comment by Gingeropolous** [comment]
@cryptonote-social , yeah I would hope there would be a bug bounty for whatever comes out of the effort. I mean, thats sort of the benefit of implementing it on a working network - hashrate spikes can be observed, and optimizations get out there - see XMR.

**Comment by Gingeropolous** [comment]
@tevador , can you comment on @hyc 's comment:

https://github.com/monero-project/monero/issues/3545#issuecomment-379979949

(unless you have already) 

specifically, 

> Why change on every nonce? Because we already know that companies like Bitmain have automatic tools to turn e.g. C code into e.g. RTL - Someone using FPGAs could easily reconfigure if you only use one algorithm per block. Generating a new program for every nonce eliminates any such advantage. It also eliminates a lot of compiler advantage too - both AOT and JIT compilation trade off compile time for run time on code that executes more than once. Since the generated code will only ever execute once, complex AOT or JIT optimization are pointless.

does your random generator do one random program per block, and then send nonces through that program?

**Comment by tevador** [comment]
Even Monero has a history of major performance improvements. See [this blog post](https://da-data.blogspot.com/2014/08/minting-money-with-monero-and-cpu.html). They wrote a 100x faster miner and at one point controlled about 60% of the network hashrate. Eventually the optimizations were published and now everyone is hashing at the improved speed.

I don't think anyone will be able to find a 100x faster javascript VM.

@Gingeropolous My algorithm works the same way as hyc's in this regard. The hash of the block header serves as the seed for the random generator. This means that any change in the block header (different nonce, different merkle root hash, different previous block hash) will change the program.

**Comment by cryptonote-social** [comment]
With Cryptonote/Cryptonight, the bytecoin authors intentionally published a broken implementation they knew they could easily speed up in order to mine the vast majority of the coins (in addition to scamming folks into believing the network was active for 2 years when in fact they just pocketed 2 years worth of blocks).   So that situation is a bit .... special.

**Comment by binaryFate** [comment]
@tevador if you're interested to optimize parameters using the dedicated research tool I mentioned, I suggest that I set up an example configuration of everything required as a starting point, but I need to discuss with you the parameter specifications. 
Note that I would set everything up on Linux, as running irace on Windows is a pain. But then irace would support testing candidates in parallel (on every cores or even on a larger cluster of multiple machines) out of the box.

I think the easiest is that you reach out to me on IRC (same nick).

My interest in general is to end up with a procedure to get to a new suitable PoW that would be reproducible for Monero one day, or from which we can learn useful feedback.

**Comment by tevador** [comment]
@binaryFate I followed the PDF guide on the website and managed to get it to run. Anyways, I still have some questions about the configuration. We can discuss on IRC.

**Comment by tevador** [comment]
@binaryFate I added irace configuration files into the repository: https://github.com/tevador/RandomJS/tree/master/doc/irace

You can just clone the repo, follow the installation instructions in README to install mono and NodeJS and then build it. You have to run the sandbox with `node sandbox.js --complexity` to enable complexity measurement. This is important.

Irace should be ready to run as-is (target-runner assumes you cloned the repo into ~/RandomJS and the machine has 32 cores), I added the result of yesterday's test run as one of the inputs in configurations.txt (my manually optimized config is still better, though).

I have increased the default runtime target to 10 ms because I found that the VM runs slower when complexity measurement is enabled (probably due to increased garbage collection) . We might have to modify it depending on the performance of the machine where it will run. You can benchmark the performance by running the default configuration: `./Tevador.RandomJS.Test.exe --count 10000 --threads 32`. It will show the average runtime (I tested on Xeon E5-1660 and the average was 14 ms).

**Comment by binaryFate** [comment]
@tevador Great I'll have a look, good job on understanding how to set it up by yourself mostly :)

First comment after quick look is that the number of instances (aka seeds) must be at least in the hundreds so they do not need to repeat, currently it is testing on a tiny subset of 5 seeds over and over again.

**Comment by tevador** [comment]
> the number of instances (aka seeds) must be at least in the hundreds 

Is it really necessary, though? Each seed is actually 1000 seeds, because each run creates 1000 programs. Speaking of which, I modified the seeds so the sequences don't overlap (the difference must be at least 1000).

**Comment by krtschmr** [comment]
If we do merge mininig with monero, we can already test for Tari.
Wownero seems to be the production testnet for M00nero.

**Comment by binaryFate** [comment]
> Is it really necessary, though? Each seed is actually 1000 seeds, because each run creates 1000 programs. Speaking of which, I modified the seeds so the sequences don't overlap (the difference must be at least 1000).

Yes it is, otherwise you get 5 (at most) different measurements only for the candidates. This is way too small of a sample, and does not allow proper statistically significant discrimination between them.

**Comment by tevador** [comment]
@binaryFate I have commited a new instances file with 200 seeds.

**Comment by binaryFate** [comment]
Great. Can you merge my [PR](https://github.com/tevador/RandomJS/pull/4)? It's just details for convenience and explanations if somebody else wants to give it a try.

I am doing some trial runs now. I'll post some advise on what configuration to use soon (mostly which statistical test to use -- makes a big difference and should be inline to what you are measuring as fitness).

**Comment by tevador** [comment]
I merged the PR. Also I did a small modification in the parameters list (2 parameters removed from optimization).

**Comment by RocksteadyTC** [comment]
Hey TurtleCoin here, we're trying to get merge mining going. Is there room for us in this discussion?

**Comment by guystyles** [comment]
Y’all should talk to Aeon too. Cryptonight unite

**Comment by ghost** [comment]
Hi @RocksteadyTC, it looks the rough consensus is to not to merge mine and go for a new POW. However, the idea of merge mining may still be of interest when [if] we incorporate a second layer protocol for digital assets. Perhaps opening a new issue to discuss merge mining on cryptonote in general would be helpful for everyone.

**Comment by RocksteadyTC** [comment]
Thanks for the heads up @wowario 

We currently have a bounty to get https://github.com/turtlecoin/turtle-pool and https://github.com/turtlecoin/node-cryptonote-util working with merge mining that a few people are poking at. I'll reference this post if we have anything worth reporting. We don't share a POW with anybody, but any developments should easily translate.

**Comment by RocksteadyTC** [comment]
We currently have a bounty to get https://github.com/turtlecoin/turtle-pool and https://github.com/turtlecoin/node-cryptonote-util working with merge mining that a few people are poking at. I'll reference this post if we have anything worth reporting. We don't share a POW with anybody, but any developments should easily translate.

**Comment by zawy12** [comment]
@hyc via @Gingeropolous 

> Someone using FPGAs could easily reconfigure if you only use one algorithm per block. 

Is it feasible to optimize and re-program an FPGA for @tevador's output in only 2 minutes?

**Comment by tevador** [comment]
@zawy12 RandomJS changes the program every nonce, not every block. Using an FPGA is out of the question.

**Comment by zawy12** [comment]
@tevador What is the variation in solvetime per nonce?  Will it be possible for an ASIC to skip nonces that return a POW that seems harder to solve than others?  What if there's a certain class of problems an ASIC is 10x faster at doing, so it keeps changing the nonce until it gets that class of problem?  If that type of thing is possible, then changing only per block might be better.

**Comment by tevador** [comment]
@zawy12 The final code generator settings are still being optimized. But we are aiming for the worst runtime out of 1000 programs to be at most 5-10 times the average runtime. This should limit nonce-skipping optimizations, because it still costs some time to generate and analyze a program and if you throw away the result, your effective hashrate is decreased.

For example, if generating a program takes 0.5 ms and running it takes 7.5 ms on average and you skip 9 our of 10 nonces, you would need to reduce the average runtime under 3 ms to actually gain an advantage (10*0.5+3 = 8 ms) compared to someone who runs all programs (0.5+7.5 = 8 ms).

Also guessing if a program will be fast or slow to run based on its source code is not an easy task. Perhaps someone could use a neural network for that, but I don't think you could actually gain a significant advantage.

**Comment by binaryFate** [comment]
- Some candidates take a very very long time, like > 300s or even sometimes > 1000s (while most others take < 20s). I suggest to stop them before doing the 1000 repetitions when the average is already known to be so bad, and just extrapolate the overall results (or just return a very-bad-on-purpose value)? Would save *a lot* of time and thus we could increase budget and run the whole tuning much faster. Can you include this sort of check in your program, so if a certain time X has already passed and only N iterations have been done, we know we can give up on this candidate and do not need to do 1000 iterations?

- What about the difference of measurements on different machines? What are we optimizing for exactly? If there is a "target" time included in the fitness function, we will get a completely different result based on the CPU we are using. The tuning process will simply optimize for the particular machine it is running on. What do you think?

- I witnessed quite some variations across "instances" for the same candidate, see this one for instance (candidate \#14 that you can also see in my results, it is typically around 120-150 fitness, so I wonder how it got 64 here)
`|x|          3|         15|         14|    64.36999813|         36|00:37:38|+0.95|0.96|0.6777|`

- It seems you are measuring system time, not user (process) time. For instance after hibernating then waking up computer I get:
```
# 2018-06-09 00:10:41 CEST: DONE (45) Elapsed: 17913.75
2147483647
```
This looks like an awful fitness just because some hours passed while my computer was sleeping.

- Some parameters are defined as categorical ("c" in parameter file) while they should be ordinal ("o").
Categorical are for parameters whose values are discrete without any implicit order. Ordinal are for parameters whose values have an implicit order, like "high", "medium", "low".
See for instance MaxLoopCyclesRange_Min. Just replace the "c" with an "o" that's it.
The difference between the two is in the sampling of new candidates: if a successful candidate uses a particular value X for this parameter, its offspring will have more chance to use values close to X if the parameter is ordinal. Otherwise if the parameter is only categorical, there isn't any order assumed by the distribution.

- I believe we should use the [t-test](https://en.wikipedia.org/wiki/Student%27s_t-test) rather than the [Friedman test](https://en.wikipedia.org/wiki/Friedman_test). The Friedman test (default of irace) relies on the rank of the candidates independently of the actual values, while the t-test relies on the values themselves. 
To make it easier to decide, I propose the following question:
"Consider a candidate X who got a certain fitness, say 20, on a certain instance. And another candidate Y who got almost the same as X, specifically 20.00001. Do you think that (i) the tuning process should consider that X is clearly unambiguously better than Y on that instance, or (ii) X and Y performed almost similarly on that instance and we should not draw much conclusion from this run due to the very small difference of absolute values?". If you answer (i), you probably want the F-test. If you answer (ii), you probably want the t-test.
In my trial (see below), I used the t-test.

- I ran a trial using the following command line (and untouched files from you, so I did not correct the "o" vs "c" of parameters nor anything else):
`irace --debug-level 2 --max-experiments 500 --test-type t-test --seed 123`
(Setting an initial seed for irace typically allows to make the whole tuning reproducible. However since you measure execution times and we will get different measurements everytime, we won't have reproducibility).
The results are posted here: https://paste.fedoraproject.org/paste/esHWpPtSh~BjtxYOangOVg
For the record, the best candidate (\#51 in my run) was this one:
`--XML_AssignmentOperators_Basic 10 --XML_AssignmentOperators_Add 3 --XML_AssignmentOperators_Sub 2 --XML_AssignmentOperators_Mul 1 --XML_AssignmentOperators_Div 3 --XML_AssignmentOperators_Mod 1 --XML_AssignmentOperators_PreInc 4 --XML_AssignmentOperators_PostDec 3 --XML_AssignmentOperators_PreDec 1 --XML_AssignmentOperators_PostInc 2 --XML_UnaryOperators_Plus 1 --XML_UnaryOperators_Typeof 4 --XML_UnaryOperators_Minus 3 --XML_UnaryOperators_Not 4 --XML_UnaryOperators_Sqrt 2 --XML_UnaryOperators_Exp 3 --XML_UnaryOperators_Log 3 --XML_UnaryOperators_Sin 2 --XML_UnaryOperators_Cos 2 --XML_UnaryOperators_Atan 1 --XML_UnaryOperators_Floor 1 --XML_UnaryOperators_Ceil 1 --XML_UnaryOperators_Abs 2 --XML_UnaryOperators_Trunc 1 --XML_UnaryOperators_Cbrt 1 --XML_BinaryOperators_Add 7 --XML_BinaryOperators_Comma 2 --XML_BinaryOperators_Sub 4 --XML_BinaryOperators_Mul 1 --XML_BinaryOperators_Div 2 --XML_BinaryOperators_Mod 2 --XML_BinaryOperators_Less 2 --XML_BinaryOperators_Greater 2 --XML_BinaryOperators_Equal 3 --XML_BinaryOperators_NotEqual 3 --XML_BinaryOperators_Min 1 --XML_BinaryOperators_Max 2 --XML_BinaryOperators_BitAnd 2 --XML_BinaryOperators_BitOr 3 --XML_BinaryOperators_Xor 2 --XML_Literals_String 2 --XML_Literals_Numeric 3 --XML_Literals_Object 2 --XML_NumericLiterals_Boolean 3 --XML_NumericLiterals_DecimalInteger 2 --XML_NumericLiterals_BinaryInteger 1 --XML_NumericLiterals_OctalInteger 4 --XML_NumericLiterals_SmallInteger 3 --XML_NumericLiterals_HexInteger 2 --XML_NumericLiterals_FixedFloat 2 --XML_NumericLiterals_ExpFloat 4 --XML_Expressions_AssignmentExpression 5 --XML_Expressions_BinaryExpression 4 --XML_Expressions_FunctionExpression 5 --XML_Expressions_Literal 2 --XML_Expressions_TernaryExpression 2 --XML_Expressions_UnaryExpression 2 --XML_Expressions_VariableInvocationExpression 2 --XML_Expressions_FunctionInvocationExpression 1 --XML_Expressions_VariableExpression 3 --XML_Expressions_EvalExpression 2 --XML_Expressions_ObjectSetExpression 2 --XML_Expressions_ObjectConstructorExpression 2 --XML_Statements_ReturnStatement 2 --XML_Statements_BreakStatement 3 --XML_Statements_AssignmentStatement 7 --XML_Statements_IfElseStatement 9 --XML_Statements_ForLoopStatement 3 --XML_Statements_BlockStatement 5 --XML_Statements_VariableInvocationStatement 5 --XML_Statements_ObjectSetStatement 3 --XML_AssignmentInForLoop_Add 1 --XML_AssignmentInForLoop_Sub 4 --XML_AssignmentInForLoop_Mul 3 --XML_AssignmentInForLoop_Div 3 --XML_AssignmentInForLoop_PostInc 1 --XML_AssignmentOperators_Div 2 --XML_GlobalVariablesCountRange_Min 6 --XML_GlobalVariablesCountRange_Span 0 --XML_LocalVariablesCountRange_Min 0 --XML_LocalVariablesCountRange_Span 3 --XML_BlockStatementsRange_Min 16 --XML_BlockStatementsRange_Span 1 --XML_FunctionParametersCountRange_Min 1 --XML_FunctionParametersCountRange_Span 3 --XML_StringLiteralLengthRange_Span 15 --XML_MaxCallDepthRange_Min 3 --XML_MaxCallDepthRange_Span 1 --XML_MaxLoopCyclesRange_Min 3000 --XML_MaxLoopCyclesRange_Span 3000 --XML_MaxStringLengthRange_Min 35 --XML_MaxStringLengthRange_Span 15 --XML_ObjectLiteralSizeRange_Min 0 --XML_ObjectLiteralSizeRange_Span 4 --XML_EvalStringLength_Min 8 --XML_EvalStringLength_Span 1 --XML_ObjectSetPropertyCount 4 --XML_VariableInitializerDepth 1 --XML_MaxFunctionDepth 1 --XML_MaxObjectLiteralDepth 1 --XML_MaxExpressionDepth 3 --XML_MaxStatementDepth 3 --XML_ConstVariableChance 0 --XML_ElseChance 0.29 --XML_AllowFunctionOverwriting false --XML_VariableSelectorScopeFactor 2 --XML_ForLoopVariableBoundsChance 0.92 --XML_AllowFunctionInvocationInLoop false`

**Comment by tevador** [comment]
> Some candidates take a very very long time, like > 300s or even sometimes > 1000s (while most others take < 20s). I suggest to stop them before doing the 1000 repetitions when the average is already known to be so bad, and just extrapolate the overall results (or just return a very-bad-on-purpose value)? Would save a lot of time and thus we could increase budget and run the whole tuning much faster. Can you include this sort of check in your program, so if a certain time X has already passed and only N iterations have been done, we know we can give up on this candidate and do not need to do 1000 iterations?

OK. I will add a time limit for objective function evaluation and then we can run the optimization with higher budget.

> What about the difference of measurements on different machines? What are we optimizing for exactly? If there is a "target" time included in the fitness function, we will get a completely different result based on the CPU we are using. The tuning process will simply optimize for the particular machine it is running on. What do you think?

You have to benchmark your machine first and then adjust the runtime target as I wrote above.
Run `./Tevador.RandomJS.Test.exe --count 10000 --threads 32` and observe the average runtime value. Mine was 14 ms. Default target is 10 ms. Adjust using command line option `--runtimeTarget 0.008` (example for 8 ms target if your machine is 25% faster).

>I witnessed quite some variations across "instances" for the same candidate

During my tests it seemed pretty consistent. Hopefully irace can handle it.

> It seems you are measuring system time, not user (process) time. For instance after hibernating then waking up computer I get:

I use the [process.hrtime](https://nodejs.org/docs/latest-v8.x/api/process.html#process_process_hrtime_time) function, which should measure CPU time correctly. The elapsed time you are showing is not from my program (perhaps it's the output from irace?).

> Some parameters are defined as categorical ("c" in parameter file) while they should be ordinal ("o").

OK. I didn't know about that. I will change it to "o".

> "Consider a candidate X who got a certain fitness, say 20, on a certain instance. And another candidate Y who got almost the same as X, specifically 20.00001. Do you think that (i) the tuning process should consider that X is clearly unambiguously better than Y on that instance, or (ii) X and Y performed almost similarly on that instance and we should not draw much conclusion from this run due to the very small difference of absolute values?". If you answer (i), you probably want the F-test. If you answer (ii), you probably want the t-test.

Definitely t-test.

**Comment by binaryFate** [comment]
> You have to benchmark your machine first and then adjust the runtime target as I wrote above.
Run ./Tevador.RandomJS.Test.exe --count 10000 --threads 32 and observe the average runtime value. Mine was 14 ms. Default target is 10 ms. Adjust using command line option --runtimeTarget 0.008 (example for 8 ms target if your machine is 25% faster).

Ok, I missed that. Makes sense as an approach.
What about the 32 threads though? For instance on my laptop I have 4 cores. Does the result depends on the number of threads? (Ideally I suppose we don't want it to depend on that).

> I use the process.hrtime function, which should measure CPU time correctly. The elapsed time you are showing is not from my program (perhaps it's the output from irace?).

Yes the "elapsed time" output is from irace, this run gave a very very bad fitness (2147483647, which is 2^31-1) so I have a doubt if you are using the elapsed time somehow. 

I am not familiar with nodejs, but from the doc of process.hrtime it measures _real time_. Real time is not what you want, as if your process sleeps for any amount of time, or the system is busy doing other things, you will get wrong results.

**Comment by tevador** [comment]
> What about the 32 threads though? For instance on my laptop I have 4 cores. Does the result depends on the number of threads? (Ideally I suppose we don't want it to depend on that).

Adjust the number of threads according to the number of cores of your CPU (I assumed you were using the 32 core machine you told me about). By the way, laptops are not the best machines for this task because the mobile CPUs usually have a high turbo frequency, which means the CPU speed is not constant and it adds variation for the runtimes. I used a machine with disabled turbo.

> Yes the "elapsed time" output is from irace, this run gave a very very bad fitness (2147483647, which is 2^31-1) so I have a doubt if you are using the elapsed time somehow.
> 
> I am not familiar with nodejs, but from the doc of process.hrtime it measures real time. Real time is not what you want, as if your process sleeps for any amount of time, or the system is busy doing other things, you will get wrong results.

`2147483647` is output only if one of the test programs crashes. In this case, the runtime is not used in the calculation at all.

I'm not aware of any better method of measuring runtime in nodejs. I don't think it's a problem unless you hibernate the computer in the middle of the optimization or run some other demanding task at the same time.

**Comment by binaryFate** [comment]
> Adjust the number of threads according to the number of cores of your CPU (I assumed you were using the 32 core machine you told me about). By the way, laptops are not the best machines for this task because the mobile CPUs usually have a high turbo frequency, which means the CPU speed is not constant and it adds variation for the runtimes. I used a machine with disabled turbo.

I was just playing around to find a suitable setup. We're getting there :)


> 2147483647 is output only if one of the test programs crashes. In this case, the runtime is not used in the calculation at all. 

Ok

> I'm not aware of any better method of measuring runtime in nodejs. I  don't think it's a problem unless you hibernate the computer in the  middle of the optimization or run some other demanding task at the same  time.

It's really not ideal especially with a high-level interpreted language, but after some googling there does not seem to be alternatives, indeed. If anyone knows how to measure process time (not real time) with nodejs please chime in.

@tevador Let me know when you'll have updated the sources to kick out bad candidates faster as discussed (maybe also update the scenario file to use the t-test so the configuration files reflect our discussions). I think the setup will be ok then and we can launch a full tuning.

**Comment by tevador** [comment]
@binaryFate I have pushed the changes:
* There is a new `--timeout` option to cap the evaluation time (returns 2147483647 if timeout is exceeded). I set the timeout to 120 seconds in target-runner.
* New objective function (now also tests EvalExpression for SyntaxErrors).
* Corrected irace parameters (t-test, ordinal parameters)

We will optimize for a 3.3 GHz Intel Sandy Bridge CPU. You have to run the benchamark:
1. Make sure to use the latest `ProgramOptions.xml`
1. Run `node sandbox.js --complexity`
1. Run `./Tevador.RandomJS.Test.exe --threads N --evalTest` where `N` is the number of threads you will use for optimization
1. Find "Runtime Avg". On the benchmark machine it's approximately 0.00105, so it almost matches the target of 10 ms. Adjust your target with the `--runtimeTarget` option.
1. You can also adjust the `--timeout` option, for example to 4 times the total runtime of the benchmark (on the benchmark machine it was `Completed in 29.8903988 seconds` with 6 cores).

**Comment by tevador** [comment]
I have pushed some changes including about 30% speed improvement for objective evaluation and a new configurable parameter for the generator.

**Comment by binaryFate** [comment]
@tevador thanks for the updates!

Should I launch a larger optimization on big machine now or are you working on something at the moment?

**Comment by tevador** [comment]
@binaryFate You can try to run a larger optimization with a budget of ~5000 (shouldn't take more than 1 day with 32 cores). Let's see what the results will be. 

The newly added parameter `FunctionValueOfOverride` is meant to decrease the amount of dead code in the program (functions that are never called), but it slows the program down, so I'm expecting the optimizer will set it to `false`. If that happens, we can run a second optimization with this parameter fixed at `true` and compare the results. The objective function doesn't account for dead code, unfortunately.

I'm not making any changes in RandomJS at the moment. I'm trying to fix bugs in the XS engine so we can use it instead of the V8.

**Comment by hyc** [comment]
> I'm trying to fix bugs in the XS engine so we can use it instead of the V8.

This worries me. Another good reason to stick to ECMA5 spec instead of ECMA6.

**Comment by tevador** [comment]
@hyc With ES5, we would lose:
* Block-scoped variables
* ... parameter (for vararg functions)
* Constants
* Lexical this (arrow functions)
* Binary and octal literals
* Math.trunc and Math.cbrt
* String templates

I listed only features used by RandomJS (in descending order how easy they would be to remove). Most of these features contribute to the complexity of the program.

Even if we sacrifice all these features, there is no guarantee ES5 engines will work without issues.

Current state of the XS engine:
* \>97% of programs match output with the V8
* More than half of the differences are caused by the EvalExpression. Seems like for a given string, one engine can produce a SyntaxError and the other engine produces a ReferenceError (or vice versa). This probably can't be considered a bug because the order of errors is not specified. I will most likely need to modify the generator or accept one engine as the reference.
* About 1% are numeric errors:
    * One bug in XS which involves integer overflow.
    * Some cases of different rounding, for example  `17876395023154640000/-13528.07885` rounds to `-1321428949473830.2` in V8 and `-1321428949473830.3` in XS. ES doesn't specify rounding or precison for Number.toString, so technically both are correct (although rounding to even is the standard). I think the XS could be modified to round to even.

The rounding issue and SyntaxError/ReferenceError differences might also affect ES5 engines.

**Comment by hyc** [comment]
OK. If you believe the differences can be addressed, then go for it.

PS, did you see this compatibility matrix? http://kangax.github.io/compat-table/es6/

**Comment by tevador** [comment]
@hyc Yes, I was looking for alternative engines in this table. Looks like XS is the only 'small' ES6 engine.

Btw here is my forked XS engine: https://github.com/tevador/moddable
I could use some help fixing the numeric issues.

**Comment by tevador** [comment]
I have fixed the numeric bugs in XS. Apart from the different treatment of SyntaxError and ReferenceError, the engines match for 99.9% of programs. There is still a few [minor issues](https://github.com/tevador/moddable/issues) to be resolved.

I'm quite satisfied with the complexity of the programs generated by RandomJS. The amount of recursion, indirection and aliasing is overwhelming when doing manual analysis of program outputs.

Interestingly, I have also rediscovered a [bug](https://bugs.chromium.org/p/v8/issues/detail?id=896) in V8, when the engine treats invalid regular expression as a runtime error rather than parsing error.

**Comment by jwinterm** [comment]
What's the benefit to using XS as a reference rather than google's V8 engine? Is it a licensing issue?

**Comment by Gingeropolous** [comment]
i asked the same elsewhere, according to @hyc , the idea was to find a smaller and simpler engine for the reference implementation @jwinterm . v8 is big and bloated.

**Comment by tevador** [comment]
@jwinterm 
XS ~ 87 000 LoC
V8 ~ 1 870 000 LoC

**Comment by tevador** [comment]
I have finally tested the performance of XS. When running in a separate process while comunicating via stdin/stdout, XS is only about 1% slower (almost within margin of error) than the V8 (represented by a [minimal wrapper](https://github.com/tevador/RandomJS/blob/master/src/fast-eval.js) around `eval` in NodeJS).

The only big difference is memory consumption. When executing 10000 programs in sequence, the XS consumes up to 470 MiB, while the node process stays below ~80 MiB. This may be just garbage collector settings, though.

Maybe someone could squeeze a few additional percent of performance by linking libv8 directly. I think the main reason why there is very little difference between interpreted and compiled engine is that each program is executed just once. 

If nobody has any additional comments about the [generator](https://github.com/tevador/RandomJS/blob/master/doc/generator.md), I will start porting it to C++.

**Comment by tevador** [comment]
Before porting to C++, I decided to make some modifications to the generator.

It's mostly finished, I'm still running tests. These are the major changes:
* Added support for `try/catch/finally` and `throw`
* Added custom class `RandomError`
* Added a global variable to count the number of thrown errors other than `SyntaxError` (this prevents a potential optimization of the `EvalExpression` by an attacker)
* Added a global variable to count the number of function calls
* Reworked the structure of `FunctionExpression`
* Fixed nested statements not being generated correctly
* Properties of object literals can be generated from a variable

Can someone review the new documentation? https://github.com/tevador/RandomJS/blob/dev/doc/generator.md

**Comment by tevador** [comment]
One more note: the change to prevent Eval optimization causes the output of XS and V8 to differ in ~35% of cases due to differences when SyntaxError and ReferenceError are emitted.

This means that web mining in browser will be still possible, but it will produce significant amount of invalid hashes. Using the V8 for serious mining will require some modifications of the engine, which might not be easy task considering the size of the engine's codebase.

Whether it is a good or bad thing is up for discussion.

**Comment by tevador** [comment]
@moneromooo-monero @jwinterm @hyc @Gingeropolous @SChernykh Can you please review and comment on the latest changes (my two comments above)?

**Comment by SChernykh** [comment]
@tevador How big are the differences between SyntaxError and ReferenceError in XS and V8? Can they be fixed by a regexp replacement or some more sophisticated but fast code?

**Comment by tevador** [comment]
I think all the differences can be attributed to non-adherence to the specs by one or the other engine. 

Most of the differences are these two categories:

1)
```
eval('/|+/+a/=ee')
XS: SyntaxError: invalid regular expression
V8: ReferenceError: Invalid left-hand side in assignment
```
Here the regular expression is invalid (the `|` operator needs a left side). XS correctly throws a `SyntaxError`, but V8 skips the RegExp validation and throws on the assignment instead. This is a violation of the specs by the V8. I found this bug on the chromium bug tracker and they are keeping this "feature" for legacy reasons.

2)
```
eval('++1a=+2e|1')
XS: ReferenceError: no reference
V8: SyntaxError: Invalid or unexpected token
```
Here the XS thinks we are incrementing the literal `1`, so it throws a `ReferenceError`. The V8 considers the token `1a` as invalid literal and throws a `SyntaxError`. I think the XS is wrong in this case as the specs explicitly forbid that an identifier follows a numeric literal without a space: http://www.ecma-international.org/ecma-262/6.0/#sec-literals-numeric-literals

Not sure how easy it would be to fix these bugs.

**Comment by SChernykh** [comment]
I meant that maybe a simple set of replacement rules can reduce error rate from 35% to almost 0%.

**Comment by tevador** [comment]
How would such rules work exactly? If you mean replacing incorrect error by the correct one, then you need to parse the expression correctly, which means fixing the parser bugs. I don't think there is a shortcut available.

**Comment by SChernykh** [comment]
> This means that web mining in browser will be still possible, but it will produce significant amount of invalid hashes. Using the V8 for serious mining will require some modifications of the engine, which might not be easy task considering the size of the engine's codebase.

I don't think that fixing V8 would be that hard, especially if they keep some bugs "for compatibility". Do all straight-forward fixes, add some replacement rules for remaining differences from XS, and error rate will be much less than 35%.

**Comment by tevador** [comment]
Apart from non-adherence to specs, there is some gray area, apparently.

For example:
```
eval('++-b');
XS: SyntaxError: missing expression
V8: ReferenceError: Invalid left-hand side expression in prefix operation
SpiderMonkey (Firefox): SyntaxError: expected expression, got '-'
```
Here the V8 seems to group the `-` and the `b` as one expression, while the other engines find two prefix operators invalid in sequence without parentheses.

> add some replacement rules for remaining differences from XS

This may be easier said than done. I think it would require extensive changes in the parser to detect errors in the same way as the XS.
Another option would be to have a separate XS-like parser just for eval to throw early errors and if the expression is validated, it could be handed over to the V8 parser.

In any case, I don't see any issue with making the XS engine part of the PoW specification, but there should be consensus about that.

**Comment by SChernykh** [comment]
I'm all for taking XS as a reference, V8 is too monstrous for this.

> Another option would be to have a separate XS-like parser just for eval to throw early errors and if the expression is validated, it could be handed over to the V8 parser.

Yes, that's the way real miners will choose if V8 is significantly faster.

**Comment by jwinterm** [comment]
@tevador briefly looks good. I'm traveling this week so I'll try to spend some time at airport tomorrow looking over it more thoroughly and treating latest version in Windows laptop.

**Comment by tevador** [comment]
@jwinterm OK. The latest code is in the dev branch.

If you want to inspect the generated Javascript, you should use something like http://jsbeautifier.org/ because the generated code has no line breaks or indentation.

**Comment by JohannesLau** [comment]
I definitely think switching to RandomJS in the next hardfork is a great idea, it is probably the best algo for asic resistance at the moment, and wownero would be a perfect testing platform for it before potential gets implemented in monero

**Comment by tevador** [comment]
Status update: I have ported the JS generator into C++. 

What is left to do is some optimization, implementation of IPC between the generator and the XS engine (already finished on the XS side) and then the cryptography, which should be pretty straightforward (standard Blake2b hash).

**Comment by tevador** [comment]
I need some feedback: https://github.com/tevador/RandomJS/issues/5

**Comment by SChernykh** [comment]
@tevador Just out of curiosity - when is it going live on Wownero network? If everything goes smooth?

**Comment by tevador** [comment]
If Wownero has a testnet, it could be deployed there very soon (I think next week). All that is needed is to write a replacement `cn_slow_hash` function that will use randomjs instead of cryptonight.

Before we go live on mainnet, someone needs to review the implementation (not just compile and test, but actually review the whole codebase).

Also currently there is [a bug](https://github.com/tevador/moddable/issues/8) in the XS engine which causes about 1 in every 5000 programs to throw an error. I wrote a bug report in the upstream repository, but it hasn't been fixed yet. We will probably have to fix it ourselves.

Anyways, randomjs can handle JS engine crash and will restart it and reject the block, so it's not major issue.

**Comment by binaryFate** [comment]
I can't run any sucessful run with the latest version (C#), all of them fail with:

```System.AggregateException: One or more errors occurred. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at Tevador.RandomJS.ProgramFactory.GenBlock (Tevador.RandomJS.Statements.Block block, Tevador.RandomJS.Interval statementsInterval, Int32 maxDepth, StatementType list) <0x407c3a90 + 0x0003b> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenFunctionBody (Tevador.RandomJS.Expressions.FunctionExpression parent) <0x407c2190 + 0x0024b> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenFunctionExpression (IScope scope) <0x407bdb00 + 0x00183> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenExpression (IScope scope, Int32 maxDepth, ExpressionType list) <0x407bae20 + 0x002df> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenVariable (IScope scope, Boolean isParameter, Boolean isLoopCounter, Boolean isConstant, Boolean initialize) <0x407b86c0 + 0x001b7> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenProgram (System.Byte[] seed) <0x407b5e70 + 0x0014b> in <filename unknown>:0 
  at Tevador.RandomJS.Test.ParallelRunner._run (CancellationToken token) <0x407abc90 + 0x0010f> in <filename unknown>:0 
  at Tevador.RandomJS.Test.ParallelRunner+<Run>c__AnonStorey0.<>m__0 () <0x407abb40 + 0x00027> in <filename unknown>:0 
  at System.Threading.Tasks.Task.InnerInvoke () <0x7fd17f023790 + 0x0004f> in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () <0x7fd17f023050 + 0x00055> in <filename unknown>:0 
  --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at Tevador.RandomJS.ProgramFactory.GenBlock (Tevador.RandomJS.Statements.Block block, Tevador.RandomJS.Interval statementsInterval, Int32 maxDepth, StatementType list) <0x407c3a90 + 0x0003b> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenFunctionBody (Tevador.RandomJS.Expressions.FunctionExpression parent) <0x407c2190 + 0x0024b> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenFunctionExpression (IScope scope) <0x407bdb00 + 0x00183> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenExpression (IScope scope, Int32 maxDepth, ExpressionType list) <0x407bae20 + 0x002df> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenVariable (IScope scope, Boolean isParameter, Boolean isLoopCounter, Boolean isConstant, Boolean initialize) <0x407b86c0 + 0x001b7> in <filename unknown>:0 
  at Tevador.RandomJS.ProgramFactory.GenProgram (System.Byte[] seed) <0x407b5e70 + 0x0014b> in <filename unknown>:0 
  at Tevador.RandomJS.Test.ParallelRunner._run (CancellationToken token) <0x407abc90 + 0x0010f> in <filename unknown>:0 
  at Tevador.RandomJS.Test.ParallelRunner+<Run>c__AnonStorey0.<>m__0 () <0x407abb40 + 0x00027> in <filename unknown>:0 
  at System.Threading.Tasks.Task.InnerInvoke () <0x7fd17f023790 + 0x0004f> in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () <0x7fd17f023050 + 0x00055> in <filename unknown>:0 <---

Completed: 0.004
2147483647
```

All results therefore are always 2147483647

@tevador ^ sorry to post debugging stuff here, but I don't find you on IRC and don't know another way to get in touch

**Comment by tevador** [comment]
@binaryFate That's because the newly introduced options were missing in the list of irace parameters. I have pushed an updated (but untested) version of the irace configuration files if you want to try it.

Keep in mind that the C# generator with NodeJS might give slightly different results than the C++ generator and the XS engine.

The manually optimized options I'm using in the C++ version seem to match the runtime targets (some CPUs are processing even more than 100 programs per second per core).

**Comment by binaryFate** [comment]
So is the C++ version usable for optimization? I am not sure because of the few bugs/open things you mentioned like IPC and 1-out-of-5000 error thrown

**Comment by tevador** [comment]
The C++ generator needs to be recompiled every time ProgramOptions are changed, so it's unusuable for optimization. It's intended for production use only.

I think the 1-out-of-5000 error could be prevented by removing the `=` character from EvalChars. This would prevent assignment in eval, but we would need to find a suitable replacement character to keep the rate of other-than-syntax errors about the same (not sure if this is possible).

If this is done, then you can optimize with the C# generator + the XS engine. The program generation takes only about 6% of the time of one hash, so generator performance doesn't make a big difference anyways.

**Comment by jwinterm** [comment]
Lengthy discussion between gmaxwell, andytoshi, and @hyc with respect to the merit of this approach:
https://pastebin.com/spug4a8x

**Comment by tevador** [comment]
Since RandomJS has become obsolete, is there interest to implement RandomX for Wownero?

**Comment by SChernykh** [comment]
> wowario (WOW)11/12/2018
Yeah... Long term, we will eventually fork to something like RandomX

Wowario in Discord on November 12th.

Edit: there is also https://funding.wownero.com/proposal/33

**Comment by jwinterm** [comment]
@tevador I think so, possibly as soon as whenever you feel like it's ready. Personally I am in favor of another route for Monero, but I think a CPU only algorithm is probably the best route for a small coin like Wownero, and if it keeps us in sync with Monero going forward (for possible merge mining), even better imo.

How close go you reckon RandomX is to being deployable at this point?

**Comment by SChernykh** [comment]
It's not CPU-only, GPU miners will be available too.

**Comment by tevador** [comment]
@jwinterm It's 99% finalized, just one little tweak is missing, which I will try to implement this weekend. Afterwards, the plan is to:

1. Publish a complete documentation (hopefully this weekend)
1. Daemon integration
1. Pool software
1. Mining software 

For pools, they will have to be slightly modified to also send the dataset seed, which is tied to the blockchain. The CPU miner for x86 is already finished and @SChernykh volunteered to prepare a GPU miner (CUDA). If someone wants to help with the pool implementation, it could speed things up.

**Comment by jwinterm** [comment]
OK, well, I think I'd be interested in giving it a try. I will bring it up in the various other forums. I may be able to help a bit with the pool modification, and maybe @wowario with daemon integration. 

Was there a change that made GPU mining more feasible? I thought that CPUs would be superior at least in terms of hash/W based on early estimations.

**Comment by Gingeropolous** [comment]
> Was there a change that made GPU mining more feasible? 

no change... someone actually just attempted to test how well it would perform on GPUs

https://github.com/tevador/RandomX/issues/24#issue-411589379

Its not an exact test, but its a rough estimate of how well they will perform. 

> I thought that CPUs would be superior at least in terms of hash/W based on early estimations.

I think they may still be superior, but I don't know at what factor.
